portal-kernel/src/com/liferay/portal/kernel/url/URLBuilder.java

```diff
@@ -0,0 +1,184 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.kernel.url;
+
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.util.HttpComponentsUtil;
+import com.liferay.portal.kernel.util.StringBundler;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * @author Julius Lee
+ */
+public class URLBuilder {
+
+	public static URLBuilder create(String url) {
+		return new URLBuilder(url);
+	}
+
+	public URLBuilder addParameter(String name, boolean value) {
+		addParameter(name, String.valueOf(value));
+
+		return this;
+	}
+
+	public URLBuilder addParameter(String name, double value) {
+		addParameter(name, String.valueOf(value));
+
+		return this;
+	}
+
+	public URLBuilder addParameter(String name, int value) {
+		addParameter(name, String.valueOf(value));
+
+		return this;
+	}
+
+	public URLBuilder addParameter(String name, long value) {
+		addParameter(name, String.valueOf(value));
+
+		return this;
+	}
+
+	public URLBuilder addParameter(String name, short value) {
+		addParameter(name, String.valueOf(value));
+
+		return this;
+	}
+
+	public URLBuilder addParameter(String name, String value) {
+		_urlOperations.add(new URLOperation(OperationType.ADD, name, value));
+
+		return this;
+	}
+
+	public String build() {
+		if (_url == null) {
+			return null;
+		}
+
+		while (!_urlOperations.isEmpty()) {
+			URLOperation urlOperation = _urlOperations.remove(0);
```

Comment:
@julschong 
The remove(0) here means you want to remove the first one. Removing the first one is usually a key usage of a queue, it is not a general usage for List.

More importantly, removing while iterating is dangerous,  it is not always safe to do, please see this for more details: https://www.techiedelight.com/remove-elements-list-within-loop/

Please make this a normal for loop, apply the operations, then clear the list.

Reply:
got it